remove a condition that table size must be passed as power of 2

narimiran opened the pull request:
This has been brought up in the forum ([link](https://forum.nim-lang.org/t/6514#40243)):

> The only correct, portable way to specify a table size is with `rightSize`, and IMO, that should be done inside `initTable` rather than expecting users to know implementation details. Users only know how many things they want to put in a table.


`defaultInitialSize` was changed from `64` to `32` because `rightSize(32) == 64`, so the default size stays the same as before.

narimiran committed:
remove a condition that table size must be passed as power of 2

timotheecour commented on lib/pure/collections/tableimpl.nim:
```diff
@@ -124,9 +124,9 @@ template initImpl(result: typed, size: int) =
     init(result, size)
   else:
-    assert isPowerOfTwo(size)
+    var correctSize = rightSize(size)
```
* can you also change other code that uses ` assert isPowerOfTwo(initialSize)`, eg in sharedtables + other files?
* it's odd that `rightSize` is used only in 1 branch but not the other (`when ctAnd(declared(SharedTable), type(result) is SharedTable):`)

narimiran replied:
> eg in sharedtables + other files

What other files?

narimiran replied:
> used only in 1 branch but not the other

When is that other branch used, btw?

I've now quickly skimmed through `sharedtables.nim` and I don't see any calls of `initImpl` :/

timotheecour replied:
> What other files?

```
lib/pure/collections/deques.nim:70:10:  assert isPowerOfTwo(initialSize)
lib/pure/collections/setimpl.nim:19:10:  assert isPowerOfTwo(size)
lib/pure/collections/sharedtables.nim:218:10:  assert isPowerOfTwo(initialSize)
lib/pure/collections/tableimpl.nim:127:12:    assert isPowerOfTwo(size)
```

timotheecour replied:
> I've now quickly skimmed through sharedtables.nim and I don't see any calls of initImpl :/

it may be indirect; it was definitely being used when i edited that part of the code; in any case if it's not being used, a `static: doAssert false` would make sure we're not going there by accident in some future update

timotheecour replied:
`nim r tests/stdlib/tsharedtable.nim` would indeed crash when using `static: doAssert false` here
(but you've fixed the code so it should now be correct)

ee7 commented:
This seems like a good ergonomic improvement. But can we clarify the pros and cons of the current approach? As I understand:

Pros:
- Users don't need to call `rightSize` or `nextPowerOfTwo` themselves.
- Previously, code that doesn't compile in debug mode or with `-d:release` could hang forever with `-d:danger`.

Cons:
- The current approach doubles the memory usage of most previously written tables, right? It was previously required to pass a power of 2 as the `initialSize` (often using `nextPowerOfTwo` or `rightSize`) and with this PR, `rightSize` is internally called unconditionally.
- In particular, old code that allocates a table that's large relative to the user's memory could cause an out-of-memory error with the next Nim release.
- It's less ergonomic when a user wants to specify exactly how much memory a Table should occupy.

To quote @c-blake in the linked forum thread
> As the person who added `rightSize`, I agree with those points actually. The only slight gotcha is that doing as you suggest would result in old code that calls it with (the correct power of two to avoid a resize, or maybe already a `rightSize`) would allocate the _next higher_ power, wasting space. That might be a price worth paying ease-of-use wise.

> We _could_ only call `rightSize` if the argument is not a power of two, but it's probably simpler to just always use `rightSize` and just tell people. Wasting space isn't quite a "failure".

Questions:
1. Is this worth a changelog entry?
2. Was the approach of "only call `rightSize` if the argument is not a power of two" considered? (I don't have a strong opinion).
3. Should the same change be made in `sharedtables`? `sets`? `deques`? The issue that seems to have motivated this change also occurs for `sharedtables` and `sets`. See below code:

```Nim
const n = 7

when defined(tables):
  import tables

  var t = initTable[int, int](n) # Compiles with `-d:danger`
  for i in 1..2:
    t[i] = i # Hangs
  echo t.len


when defined(st):
  import sharedtables

  var st: SharedTable[int, int]
  init(st, n) # Compiles with `-d:danger`
  for i in 1..2:
    st[i] = i # Hangs
  echo st.len


when defined(sets):
  import sets

  var s1: HashSet[int]
  init(s1, n) # Compiles with `-d:danger`
  for i in 1..2:
    s1.incl(i) # Hangs
  echo s1.len

  var s2: OrderedSet[int]
  init(s2, n) # Compiles with `-d:danger`
  for i in 1..2:
    s2.incl(i) # Hangs
  echo s2.len


when defined(deques):
  import deques

  var d = initDeque[int](n) # Compiles with `-d:danger`
  for i in 1..2:
    d.addFirst(i) # Doesn't hang here.
  echo d.len
```

ee7 commented on lib/pure/collections/tableimpl.nim:
```diff
@@ -124,9 +124,9 @@ template initImpl(result: typed, size: int) =
     init(result, size)
   else:
-    assert isPowerOfTwo(size)
+    var correctSize = rightSize(size)
```
```suggestion
    let correctSize = rightSize(size)
```

timotheecour commented:
* this PR should link to https://github.com/nim-lang/Nim/issues/14895

> Was the approach of "only call rightSize if the argument is not a power of two" considered? (I don't have a strong opinion).

that's a really bad idea as explained here https://github.com/nim-lang/Nim/issues/14895#issuecomment-653796393; you'd end up having non-monotonic memory usage when running an algorithm with N, N+1, N+2, depending on whether N is a power of 2

narimiran commented:
> The current approach doubles the memory usage of most previously written tables, right? It was previously required to pass a power of 2 as the initialSize (often using nextPowerOfTwo or rightSize) and with this PR, rightSize is internally called unconditionally.

My first version used `nextPowerOfTwo` instead of `rightSize` as I thought that is the way to go, especially if a user wants a specific size.
Then I thought a bit more and came with an example where user wants to have e.g. 8 keys in the table. The table of size 8 won't be enough, it will be resized at some point — that's why I changed it to `rightSize`.

But now you can come up with the example of a user who had taken that already in mind, wants 8 keys, passes 16, and now the size is 32.

It seems like there is no "one true way" of doing this unambiguously :/

timotheecour commented:
> It seems like there is no "one true way" of doing this unambiguously :/

there is: it's `proc initTable*[A, B](initialSize = defaultInitialSize, expectedNum = -1): Table[A, B] =`

as I recommended in https://github.com/nim-lang/Nim/issues/14895#issuecomment-653796393
which would avoid those issues; you'd only take into account `expectedNum` if its != -1.

narimiran commented:
> The current approach doubles the memory usage of most previously written tables, right?

Based on my experience and the code I've seen in the wild, I think the most usages of `initTable` (and its siblings) were with the default value, without specifying the size. And for those cases, the memory usage would stay the same as before.

For other cases, where users have specified the size, it would be double. (But see my previous comment: there might be cases where resizing was happening before, so in the end it is again the same as before)

narimiran committed:
remove power-of-2 condition from sets and sharedtables

narimiran committed:
remove power-of-2 condition from deques

c-blake commented:
I agree that very little code in the wild pre-sizes either `HashSet` or any `Table`..sometimes even when it could via knowing some `.len`.  If you follow my idea in https://github.com/nim-lang/Nim/issues/14895 then the only code getting the 2x RAM allocation impact is explicit power-of-two init code.  I'm a bit more ambivalent about how bad non-monotonic behavior is here, but I do see how it could be confusing and think a `-d:noAutoRightSize` (or similarly named) guard is enough and less confusing (EDIT: ***if*** you want to support old explicit po2'ers).  I also think we should be more clear in the doc comments that it is a size hint to not imply some stricter than we want contract going forward.

c-blake commented:
I do think, at least for a transition period, supporting old `rightSize` users is not difficult and the right thing to do..easy by having it be `proc rightSize*(x: int): int {.inline.} = x` (possibly in its own new module as @Araq suggested, although since in the new set up people will just be deleting any use of `(sets|tables|..).rightSize` that seems less important).  You just need some other unexported `rightSizeInt` or `rightSizeInternal` or `hashRightSize` or something..probably just re-defined anywhere it's needed which is no big deal for a tiny unexported inline proc.

I certainly use the old `rightSize` in my own code, but I also don't mind deleting it promptly.  So, I'm not like super adamant about it, but it bears repeating as an idea since then the only users impacted would have had to be "very expert users" instead of merely "expert users". ;-)  And test cases that care about unresized hash order which only need a `shr 1` or `div 2` tossed in.  Mostly to @ee7's questions.  And there should be a changelog.md entry when the delta is finalized just in case.

narimiran committed:
use 'correctSize' for both branches

hashbackup commented:
The reason I didn't like the idea of keeping initialSize and adding adding a new expectedNum argument is because a) it complifies the API; b) it preserves the misleading notion that defaultSize specifies memory size:

> It's less ergonomic when a user wants to specify exactly how much memory a Table should occupy.

defaultSize specifies the number of table slots, not MB of memory.  How much memory the table will use depends on table implementation details like the type of keys & values, resize policy, load factors, etc.  

If a user wants to control exactly how much memory a table can use, that should be a separate argument, with units of MB, and could be used as either the initial size (with # of slots computed) or as a not-to-exceed limit.  That's a useful feature I use in HashBackup (not-to-exceed), but I also have the luxury of a lossy hash table.  For Nim tables, if the limit was exceeded, it would have to throw an exception.  It seems drastic, but if memory constraints are tightly specified, it probably makes more sense to blow up in a hash table resize rather than allowing a large resize to occur and then bombing in some unrelated part of the program because the hash table is using more RAM than expected.

If a new sizeLimit arg was used as the initial table size rather than not-to-exceed, then an exception would be raised during item insertion if the table's load factor is exceeded because no resize would be allowed.  This is probably a better way to go in a memory-constrained situation because a table resize from N to 1.5N needs 2.5N memory during the resize.

Araq commented:
In your memory constrained scenario you can also just use a BTree instead. And better stop using ``seq`` altogether since they don't offer a maximum size either.

ee7 commented:
Ah, I hadn't seen #14895.

> It seems like there is no "one true way" of doing this unambiguously :/

Yeah, I suppose Nim can't divine the user's intent here - we'd need to add an extra parameter, or do something like "make `rightSize` return a `distinct int`" as previously mentioned.

> `proc initTable*[A, B](initialSize = defaultInitialSize, expectedNum = -1): Table[A, B] =`

Mmm. If we're considering adding an extra parameter, instead of `expectedNum` I guess this is also a possibility:
```nim
proc initTable*[A, B](initialSize = defaultInitialSize, foo = true): Table[A, B] =
```
where `foo` being `true` means "call `rightSize` is even when `initialSize` is a power of two". Then users could write

```nim
let n = 7
var t1 = initTable[int, int](n) # new behavior: `n` don't need to be a power of two.
var t2 = initTable[int, int](n.nextPowerOfTwo, false) # to get the old behavior
```

I'm not saying that this is better or a good idea. But perhaps it fares better against these points from @hashbackup
> The reason I didn't like the idea of keeping initialSize and adding adding a new expectedNum argument is because a) it complifies the API; b) it preserves the misleading notion that defaultSize specifies memory size

I think the current state of this PR is reasonable, but I do think that users who did "the right thing" before should be told in a changelog entry at least that their table's initial memory use will double. It could be a breaking change for somebody doing (perhaps uncommon) things with big tables/sets or a small heap.

hashbackup commented:
@Araq Thanks for the tip.  I'm not doing anything memory constrained, but was more thinking more about hard realtime or embedded environments using Nim.  A hard limit on hash table size might useful for them.

I have done a btree test with this same small test case and the btreetables module.  It uses 26% less RAM (595M btree vs 806M hash table) but is also 12x slower.  So I guess it depends which is more important.  I'll probably do a forum post about it one of these days, because SQLite manages to create an in-memory btree for the same test in around 160MB of RAM, though 12x slower than the Nim btree module.  It would be nice to understand why and see if the Nim btree footprint could be decreased without a big runtime hit.

Araq commented:
Anyhow @narimiran please add a changelog entry and it's as good as it can be IMO.

narimiran committed:
prettify changelog.md and fix typos

narimiran committed:
add a changelog entry

ee7 commented:
Oh, and we should remove the now-duplicate `rightSize` calls in `sets.nim` and `tables.nim`, right? Otherwise this PR doubles the size of the underlying `seq` when using the `to` procs.

The `rightSize` calls are in these procs:
```nim
proc toHashSet*[A](keys: openArray[A]): HashSet[A] =
proc toOrderedSet*[A](keys: openArray[A]): OrderedSet[A] =

proc toTable*[A, B](pairs: openArray[(A, B)]): Table[A, B] =
proc toOrderedTable*[A, B](pairs: openArray[(A, B)]): OrderedTable[A, B] =
proc newOrderedTable*[A, B](pairs: openArray[(A, B)]): <//>OrderedTableRef[A, B] =
proc toCountTable*[A](keys: openArray[A]): CountTable[A] =
proc newCountTable*[A](keys: openArray[A]): <//>CountTableRef[A] =
```

On these lines:

```nim
lib/pure/collections/sets.nim
220:  result = initHashSet[A](rightSize(keys.len))
678:  result = initOrderedSet[A](rightSize(keys.len))
973:      var b = initHashSet[int](rightSize(a.len))
```

```nim
lib/pure/collections/tables.nim
332:  result = initTable[A, B](rightSize(pairs.len))
1294:  result = initOrderedTable[A, B](rightSize(pairs.len))
1782:  result = newOrderedTable[A, B](rightSize(pairs.len))
2242:  result = initCountTable[A](rightSize(keys.len))
2601:  result = newCountTable[A](rightSize(keys.len))
```

Or if you want deep links:
<details>
  <summary>sets.nim</summary>
  
https://github.com/nim-lang/Nim/blob/bbbe9282c0eb83857c50de02d109ac9544a6b065/lib/pure/collections/sets.nim#L203-L221
https://github.com/nim-lang/Nim/blob/bbbe9282c0eb83857c50de02d109ac9544a6b065/lib/pure/collections/sets.nim#L661-L679
https://github.com/nim-lang/Nim/blob/bbbe9282c0eb83857c50de02d109ac9544a6b065/lib/pure/collections/sets.nim#L971-L975
(the last one is inside a test)
</details>

<details>
  <summary>tables.nim</summary>

https://github.com/nim-lang/Nim/blob/bbbe9282c0eb83857c50de02d109ac9544a6b065/lib/pure/collections/tables.nim#L319-L333
https://github.com/nim-lang/Nim/blob/bbbe9282c0eb83857c50de02d109ac9544a6b065/lib/pure/collections/tables.nim#L1280-L1295
https://github.com/nim-lang/Nim/blob/bbbe9282c0eb83857c50de02d109ac9544a6b065/lib/pure/collections/tables.nim#L1768-L1783
https://github.com/nim-lang/Nim/blob/bbbe9282c0eb83857c50de02d109ac9544a6b065/lib/pure/collections/tables.nim#L2239-L2243
https://github.com/nim-lang/Nim/blob/bbbe9282c0eb83857c50de02d109ac9544a6b065/lib/pure/collections/tables.nim#L2598-L2602
</details>

c-blake commented:
As things stand, I think you get a 2x RAM expansion if one's old code **either** does a different semantics po2 **or** does a `rightSize`.  We don't need to "break" both (admittedly a *very* weak sense of the word "break", but people are seeming to care).

I think the same precise growth avoidance across a wide range of Nim versions going back 5 years is easy to make work with my public-identity-rightSize + private-realRightSize idea. I think the latter can also be done in one of the `include` files now.  I can do it if @narimiran doesn't want to do it (probably late today/tomorrow).

Also, apologies for starting my comments on the issue thread.  It looks from the timestamps like I began editing that post before @narimiran issued his PR..So, I guess he & I were both thinking about this during basically the exact same hour by coincidence. :-)

c-blake commented:
That keeping `rightSize` working means we can also defer the work @ee7 just mentioned until "whenever", as just one example.  If we do it the current way we should definitely fix `toHashSet` `toTable`, etc.  My way, the whole world out there gets to evolve it "whenever".  Unless they were doing an explicit po2 { EDIT: **and** care. :-) }.

timotheecour commented:
I still prefer 
`proc initTable*[A, B](initialSize = defaultInitialSize, expectedNum = -1): Table[A, B] =` as I recommended in https://github.com/nim-lang/Nim/issues/14895#issuecomment-653796393 which avoids performance regressions and also **allows user to choose whether to initialize by size or by expected number of elements. In some settings it makes sense to specify the memory it consumes rather than the number of elements it'll hold.**

This PR introduces subtle, hard to find performance regressions in this PR, here are just 2 examples:
* any caller of `toTable(pairs)` now calls `rightSize(rightSize(pairs.len))`, causing a 2X memory usage
* `LockFreeHash.nim:176:5:    newTable = newLFTable[K, V](oldLen*2)` (sure, `LockFreeHash.nim` is deprecated, but that's besides the point)

who knows how many other code like `toTable` exists in third party code. Authors may never notice, as nothing would give CT erros, but it can cause memory usage to be 2X as large and only affect production use.

c-blake commented:
In the old scheme, I think there is only one "good" way to pre-size which is the old `rightSize`.  Anything else and the code pre-sizing is assuming 2/3 max load which we may want to change or even parametrize.  Your two argument solution doesn't really change that need for deep impl understanding and impl constancy.

My idea of `rightSize`->identity and private `realRightSize` will keep all good old uses fitting as well as they ever did as well as any new ones as well as eliminating the need to even check if an argument is ever a power of two.  So, the asserts can go.

Admittedly, some sophisticated "growth policy object" could be done that did all the above and more.  If the goal is to really fix this forever, that would be better, but more work and debate.  If the goal is to keep things simple, I think my approach works well and supports what should be supported.

mratsim commented:
I might have missed something but why is rightSize returning the next power of two if we already have a power of 2?
This is a simple way to avoid that:

```Nim
func nextPowerOf2*(x: int): int =
  ## Returns x if x is a power of 2
  ## or the next biggest power of 2
  1 shl (fastLog2(x-1) + 1)
```

c-blake commented:
The meaning of `rightSize` is not "create a non-bug inducing argument" but return a `data` size that can hold the number of elements you fed to `rightSize` without inducing growth/rehashing.  So, we could change that to what you say (as I mentioned in the forum), but @timotheecour points out non-monotonicity could be confusing over in the issue.

Additionally, if some `.len` just happened to be a power of two "by accident" (possibly over and over again in a loop) then pre-sizing would not work..the `mustRehash` would still cause growth as you build the table. That very last before final growth is always the most expensive one, too.

hashbackup commented:
The previous documented advice was to use either rightSize or nextPowerOfTwo.  If I need to insert 16 items, I know it's a power of 2 and use 16 for initTable.  When 16 items are inserted, it will cause a resize, so will need the original 16 slots plus 32 slots for the new table = 48 slots until the original 16 slot table is garbage collected.  The new version would only require 32 slots and will be faster because the resize is avoided.  So it can be argued that the change will sometimes be better or sometimes be worse.  

I do think it would be nice if existing programs that already used rightSize didn't get doubled, but also think rightSize should be deprecated.  I proposed a solution for that and @c-blake proposed another, probably simpler solution.

Several people have commented that the existing interface allows specifying the amount of memory used.  

> In some settings it makes sense to specify the memory it consumes rather than the number of elements it'll hold.

I don't get this.  If I create a 10-slot hash table, how many bytes of memory are used?

timotheecour commented:
@mratsim
> why is rightSize returning the next power of two if we already have a power of 2?

as explained here https://github.com/nim-lang/Nim/issues/14895#issuecomment-653796393 you'd end up having non-monotonic memory usage when running an algorithm with N, N+1, N+2, depending on whether N is a power of 2; this makes no sense and will cause subtle performance bugs that you may not notice until used in production.

> non-monotonicity could be confusing over in the issue

it's not just confusing, it's error prone.

As a general rule, edge cases (and giving the same argument 2 different meanings depending on whether it's a power of 2) produces error prone software. 

> also think rightSize should be deprecated.

no, exposing what the stdlib uses makes sense, if you want to predict how much memory will be used.

>> In some settings it makes sense to specify the memory it consumes rather than the number of elements it'll hold.

> I don't get this. If I create a 10-slot hash table, how many bytes of memory are used?

before this PR, `initTable(initialSize)` would use `KeyValuePair[A, B].sizeof*initialSize`, predictably

hashbackup commented:
> before this PR, initTable(initialSize) would use KeyValuePair[A, B].sizeof*initialSize, predictably

Hmm, I think that's not quite right because there is also a hash code stored with them.  No idea how big it is, but it does support my point that determining memory size relies on current implementation details.

timotheecour commented:
> Hmm, I think that's not quite right because there is also a hash code stored with them.

it is correct; check the implementation

c-blake commented:
Well, `KeyValuePair` does include the hash code (for locality) but is not exported, actually.  So, that `sizeof` would only work in-module and is definitely still implementation internals. :-)

Explicit power of two's were also always error prone/leaking the guts of the impl too much.  If you did some benchmark that said resize should be at 50% or 75% all such use cases would be in trouble.  Explicit powers of two overpromise, underdeliver, and tie our hands.  `rightSize` was the only "safe" way to ever prevent growth if you knew how many elements you might insert.

No one has really explained any objection to `realRightSize` with `rightSize*` as identity.  I think if there is any will to alter the signature of these inits including just the semantics of initialSize to be a hint then we should move away from power of two.

If you really want to predict memory use (this seems rare), we can instead maybe add a `proc` to an instantiated Table/HashSet with the generic parameters bound called `bytesFor*(container, numElements: int)` or something..or `slotsFor*`, etc.

hashbackup commented:
In the future, someone might decide to put all the hash codes together and the key-value pairs together, to optimize cache access while searching, or make any number of changes to improve performance that could change the size of an entry.  Are all of those going to be blocked because of the way tables were initially released?

Another perhaps relevant point is that the initTable documentation doesn't say anything about memory size / usage.

timotheecour commented:
see https://github.com/nim-lang/Nim/pull/14932 for a better approach that avoid this PR's pitfalls in particular performance regressions such as `toTable(pairs)` which now calls `rightSize(rightSize(pairs.len))`, using at least 2X more memory that needed; even if `toTable` were fixed, third party code may be similarly affected (and would need to branch conditionally on nim devel vs nim <= 1.2 to avoid over-allocations).

EDIT: my PR got closed (see https://github.com/nim-lang/Nim/pull/14932#issuecomment-655325419) but is AFAIK still a valid counter-proposal and ready for review, see PR for details

narimiran committed:
fix double-call of 'right-size'

narimiran commented:
> `toTable(pairs)` which now calls `rightSize(rightSize(pairs.len))`

Thanks for reporting it, I've sent a fix.

ee7 commented:
> > toTable(pairs) which now calls rightSize(rightSize(pairs.len))

> Thanks for reporting it, I've sent a fix.

@narimiran your most recent commit doesn't remove the now-duplicated calls in `sets.nim`. Please see my comment https://github.com/nim-lang/Nim/pull/14926#issuecomment-654961174.

narimiran committed:
fix the same thing in sets.nim

narimiran commented:
> your most recent commit doesn't remove the now-duplicated calls in `sets.nim`

Thanks, fixed.

c-blake reviewed (commented):
This fixes it in the stdlib, but code in the wild, if there is any, needs an identity `rightSize` to be fixed as I've mentioned several times.

Varriount commented:
@c-blake So what you're suggesting is that the implementations move away from powers-of-two allocation, and that a "size hint" parameter is created?

c-blake commented:
If code in the wild uses `rightSize` and is not adjusted and we do not do make the public `rightSize` the identity then double-`rightSize` application of the non-identity is using 2x the RAM | cache footprint, not 2x the ***work*** which is both the risk & reality of `mustRehash`-naive power-of-two-guessing code.  See https://github.com/nim-lang/Nim/pull/14932#issuecomment-655400899 .  This isn't the end of the world.  People who want the optimal allocation can just remove the `rightSize` call.

What I am suggesting is just a simple extra identity proc and some new private call so even that small amount of code migration to avoid the small amount of run-time penalty is not necessary.  

So, just for example the two most recent sets & tables changes @narimiran just did would have been "nice but not necessary" or could have been done as clean-up whenever.  That's because though they called `rightSize`, that call would have just been a pass through and the "real" "3/2 adjustment"/inverse-mustRehash calc was done inside the init proc (by some new private call `slotsNeeded` or something) not outside it in the call.

So, under my proposal all (correct) code in the wild would just keep working as-is identically.  This may be overzealous backward compat, but it's also not hard to do.  Under my proposal, the only code impacted by this change would be power-of-two-guessing code which *was always performance risky* since `mustRehash` is an internal, opaque procedure that could change..and risky in the more important 1st order sense of performance of doing work (not, say, having cache be uncontended or being near some latency cliff on the deployment CPU).

I was going to do a PR to add that today, but opening a parallel PR might be considered rude.  It can also be done after you do this PR.  Or in this PR.  I hope the idea is clear but no one but @hashbackup and just now @Varriount has even acknowledged the idea in these conversations.

c-blake commented:
@Varriount - I had started typing that before I read your comment.  The `data` needs to stay power of two for a simple `and` mask to work.  This PR has been an ongoing attempt to re-interpret the `initialSize` parameter to mean the more user-visible max `.len` supported before an expensive growth event rather than its old "how much we allocate in `data`" meaning.  I'm not against the PR as-is.  I just think its "backward compatibility in not overallocating" story could be made stronger easily.  And after the fact, too.

Araq commented:
Sorry @narimiran, c-blake's idea is excellent, make the old ``rightSize`` the identity function and introduce a new proc ``hashes.guessSize`` or something with a better name.

ee7 commented:
@c-blake My apologies for not acknowledging your idea.

For what it's worth: I like it, as long as it is well-documented in doc comments and the changelog. As you say, I think it addresses all the raised concerns apart from the more minor "people who used `nextPowerOfTwo` when creating their set/table will have their set/table's initial memory usage doubled".

It ensures that code that uses `rightSize` runs as before, and does not introduce the complexity of an extra parameter or some `distinct int` approach.

If we don't ensure that old `rightSize` code works as before, this PR theoretically represents a breaking change - people who allocate over half of their heap when creating a table/set in Nim 1.2 would face out-of-memory errors on 1.4. Of course, that might not impact many people, but I think it's something that should be mentioned in a changelog. With @c-blake's approach, we should also explain to users how "code that used `nextPowerOfTwo` when creating a set/table" is impacted.

c-blake commented:
Even though it's (probably) an internal name and so does not matter much, given that there seems to have been some fixation on memory allocation instead of actual work done calling it `slotsNeededToAvoidResizeForElementCountOnInitialBuild(n: int): int` might be good. ;-) { Just kidding!!  `inverseMustRehash` is another idea. }

c-blake commented:
(and it should be probably be defined wherever `mustRehash` is - they should always be changed in lock step.)

c-blake commented:
Also, in case it wasn't obvious from your most recent small mistake, you will need to change that`rightSize` definition to the identity everywhere the current is defined - i.e. in both `sets.nim` and `tables.nim`..and in `sharedtables.nim` and the other `init` procs you want to call `inverseMustRehash` (the old rightSize calc) instead of that new `rightSize`.

narimiran committed:
introduce a new internal proc `slotsNeeded`

Deprecate the public proc `rightSize`, which is not needed anymore.
Now it is an identity function, allowing the old code to work
correctly and without extra allocations.

narimiran commented:
@c-blake I worked on a new commit without seeing your latest comment, but I think it should address all the stuff you mention.

> you will need to change that `rightSize` definition to the identity everywhere the current is defined - i.e. in both `sets.nim` and `tables.nim`..and in `sharedtables.nim`

`rightSize` was defined only in one place, and included in those other files. 

> other `init` procs you want to call `inverseMustRehash` (the old rightSize calc) instead of that new `rightSize`

Done. (I've named it `slotsNeeded`)

----

Although it doesn't matter, I removed all calls to `rightSize` in stdlib. (I don't like seeing deprecation warnings :))

c-blake commented:
Cool.  Looked it all over.  Seems good to me.  Name is fine.  Just a local anyway.  (Out in the wider world there is an expression "right sizing" that I did not invent that inspired that choice 5 years ago).

I'd say you just need a changelog.md entry.  I suggest something like:
```
The `initialSize` parameter for HashSet, *Table no longer need to be a power of two and its interpretation is now more strictly about the biggest `len()` such a collection supports without auto-growth.  People using `nextPowerOfTwo` should just remove the call.  `rightSize` is the identity and ok to leave, but is now deprecated.
```
But feel free to simplify or expand.

hashbackup commented:
Thanks, this fixes all of my concerns in #14895.

ee7 commented:
I'll try to summarise the discussion, in case anybody is looking at this in the future. Let me know if I missed something.

It also seems nice to summarise, given that at the time of merging this PR was:
- the most-commented open PR
- in the top 10 [most-commented PRs](https://github.com/nim-lang/Nim/pulls?q=is%3Apr+sort%3Acomments-desc), despite being open for less than 1.5 days
- in the top 20 [most-commented combined PRs and issues](https://github.com/nim-lang/Nim/issues?q=sort%3Acomments-desc) (of course, there's some in the RFCs repo too)

The discussion progressed through:
- a forum thread https://forum.nim-lang.org/t/6514#40243
- a GitHub issue https://github.com/nim-lang/Nim/issues/14895
- this PR
- @timotheecour's PR https://github.com/nim-lang/Nim/pull/14932

### Summary
The merged commit is:
- the original commit from @narimiran: call `rightSize` unconditionally in `tables.nim`
- extended to apply to `sharedtables.nim`, `sets.nim` and `deques.nim`
- removing the now-duplicate `rightSize` calls in the procs like `toHashSet` and `toTable`
- with a change from @c-blake to make the public`rightSize` the identity function
- and deprecating the public `rightSize`

The advantages of the merged commit are:
- it addresses the original concerns from @hashbackup, which were:
  - code that doesn't compile in debug mode or with `-d:release` could hang forever when compiled with `-d:danger` (see also the code at the bottom of https://github.com/nim-lang/Nim/pull/14926#issuecomment-654750599).
  - "users shouldn't be expected to know implementation details - they just know how many things they want to put in a container".
- it makes the API more ergonomic, compared to the pre-PR state that required the user to call `rightSize` or `nextPowerOfTwo`, or specify a power of two. It also has the same advantage over "just change `assert` to `doAssert` to fix the hang with `-d:danger`".
- it does not introduce the complexity of e.g.: a new proc; a new parameter with a default value (see https://github.com/nim-lang/Nim/pull/14926#issuecomment-655053214); or some `distinct int` approach.
- it avoids making powers of two a special case (e.g. "only call `rightSize` if the argument is not a power of two") and avoids non-monotonicity of memory usage (it would be strange if `initHashSet[string](31)` used more memory than `initHashSet[string](32)`).
- it ensures that code that uses `rightSize` runs as before.
- it uses an internal `rightSize` rather than an internal `nextPowerOfTwo`, avoiding the performance problem stated in https://github.com/nim-lang/Nim/pull/14932#issuecomment-655400899.

The main discussed disadvantage:
- Currently, Nim 1.4 will double the initial memory usage of a `Table` or `HashSet` that specifies `initialSize` by manually passing a power of two or using `nextPowerOfTwo`. 

or as stated by @c-blake in https://github.com/nim-lang/Nim/pull/14926#issuecomment-655414404
> So, under my proposal all (correct) code in the wild would just keep working as-is identically. This may be overzealous backward compat, but it's also not hard to do. Under my proposal, the only code impacted by this change would be power-of-two-guessing code which was always performance risky [...]

and in https://github.com/nim-lang/Nim/pull/14926#issuecomment-654773959
> the only users impacted would have had to be "very expert users" instead of merely "expert users". ;-)

### Examples
To illustrate, here are the main ways to create a `HashSet` of strings, given some integer `n` that is the expected number of items:
1. `var s = toHashSet(["foo", "bar"])`
1. `var s = initHashSet[string]()`
1. `var s = initHashSet[string](1024)`
1. `var s = initHashSet[string](n.nextPowerOfTwo)`
1. `var s = initHashSet[string](n.rightSize)`

(For clarity, let's ignore the `var s: HashSet[string]; init(s)` versions and the deprecated `initSet` and `toSet`).

This PR originally kept options 1 and 2 working as before, but theoretically represented a breaking change to code that did option 3, 4 or 5; people who allocated over half of their remaining RAM when creating a table/set in one of those ways in Nim 1.2 would face out-of-memory errors in Nim 1.4.

With the merged commit, the public `rightSize` is the identity function, which limits that impact to code that does option 3 or 4.

hashbackup commented:
One of the contentious points of this discussion was that if initTable was called with a power of 2, the old way would leave it alone while the new way now calls slotsNeeded, increasing the amount of memory used.  @c-blake responded to this concern in detail on #14932.  To further confirm that we took the right approach, here are two benchmarks on Nim 1.2.1:
* add 4096 items to a map created with nextPowerOfTwo, ie, leave it at 4096
* add 4096 items to a map created with rightSize (the new way)

First leaving the table as a power of 2 with nextPowerOfTwo(4096):
```
import tables
import math

const
  els = 4096
  loops = 1_000_000

for i in 0 .. loops:
  var map = initTable[int, int](nextPowerOfTwo(els))
  for j in 0 ..< els:
    map[j] = j

ms:nim jim$ nim c -d:danger po2table
Hint: 22720 LOC; 0.366 sec; 24.625MiB peakmem; Dangerous Release build; proj: /Users/jim/nx/po2table; out: /Users/jim/nx/po2table [SuccessX]

ms:nim jim$ /usr/bin/time -l ./po2table
       64.60 real        64.58 user         0.01 sys
   5320704  maximum resident set size
```
Now the new way, using rightSize(4096):
```
import tables

const
  els = 4096
  loops = 1_000_000

for i in 0 .. loops:
  var map = initTable[int, int](rightSize(els))
  for j in 0 ..< els:
    map[j] = j

ms:nim jim$ nim c -d:danger po2table
Hint: 22719 LOC; 0.352 sec; 24.625MiB peakmem; Dangerous Release build; proj: /Users/jim/nx/po2table; out: /Users/jim/nx/po2table [SuccessX]

ms:nim jim$ /usr/bin/time -l ./po2table
       45.22 real        45.20 user         0.02 sys
   4890624  maximum resident set size
```
There is a 43% performance penalty for leaving the table a power of 2 and instead of saving memory, it uses 9% more memory.

The discussion and summary sort of imply that this change has a negative impact if the table size was specified as a power of 2.  While that is possible, this benchmark shows that the new way can also have a very positive impact, even if the table is a power of 2.
